cluster/kubernetes/namespacer.go

```diff
@@ -0,0 +1,166 @@
+package kubernetes
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	crdv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	crd "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/discovery"
+	discocache "k8s.io/client-go/discovery/cached"
+	toolscache "k8s.io/client-go/tools/cache"
+
+	kresource "github.com/weaveworks/flux/cluster/kubernetes/resource"
+)
+
+type namespaceViaDiscovery struct {
+	fallbackNamespace string
+	disco             discovery.DiscoveryInterface
+}
+
+type namespaceDefaulter interface {
+	GetDefaultNamespace() (string, error)
+}
+
+// NewNamespacer creates an implementation of Namespacer
+func NewNamespacer(ns namespaceDefaulter, d discovery.DiscoveryInterface) (*namespaceViaDiscovery, error) {
+	fallback, err := ns.GetDefaultNamespace()
```

Comment:
Since the user can (at least theoretically) mount and modify the kubeconfig file on the fly (including the default namespace), I am not sure it's a good idea to simply read it once and be happy with it.

Reply:
Mounting a new file === creating a new pod, no?

Reply:
Oh, as in, you can modify the contents of the ConfigMap or Secret, and that eventually gets reflected in the pod's filesystem. Yes, that can happen, true. It's not assumed that containers will adapt to those changes -- but, we can fairly cheaply and helpfully do so.

Reply:
Although -- not sure I want to do an exec every time EffectiveNamespace is called, so this might involve an explicit "Refresh" method on the namespacer, or some indirection

Reply:
Happy to create a test for this after merging https://github.com/weaveworks/flux/pull/1765 and rebasing on top of it.

Reply:
On reflection, I'm not keen on putting in a Refresh method, as the abstractions stand at the minute. I think it'd be better to live with requiring a restart on changes for now, and revisit how the namespacing works afterwards.

Reply:
Agreed